// vite-plugin-babel/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var babel = require('@babel/core');
var vite = require('vite');
var fs = require('fs');
var path = require('path');

const esbuildPluginBabel = (options) => ({
  name: "babel",
  setup(build) {
    const { transformFilter, namespace = "", config = {}, loader } = options;
    const resolveLoader = (args) => {
      if (typeof loader === "function") {
        return loader(args.path);
      }
      return loader;
    };
    const transformContents = async (args, contents) => {
      const babelOptions = babel.loadOptions({
        filename: args.path,
        ...config,
        caller: {
          name: "esbuild-plugin-babel",
          supportsStaticESM: true
        }
      });
      if (!babelOptions) {
        return { contents, loader: resolveLoader(args) };
      }
      if (babelOptions.sourceMaps) {
        babelOptions.sourceFileName = path.relative(process.cwd(), args.path);
      }
      return babel.transformAsync(contents, babelOptions).then((result) => ({
        contents: result?.code ?? "",
        loader: resolveLoader(args)
      }));
    };
    build.onLoad({ filter: /.*/, namespace }, async (args) => {
      if (!transformFilter(args.path)) return;
      const contents = await fs.promises.readFile(args.path, "utf8");
      return transformContents(args, contents);
    });
  }
});

function testFilter(filter, id) {
  if (!filter) return true;
  return typeof filter === "function" ? filter(id) : filter.test(id);
}

const DEFAULT_INCLUDE = /\.jsx?$/;
const viteMajorVersion = Number(vite.version.split(".")[0]);
const babelPlugin = ({
  babelConfig = {},
  filter,
  include,
  exclude,
  apply,
  enforce = "pre",
  loader,
  optimizeOnSSR = false
} = {}) => {
  const isVite8OrHigher = viteMajorVersion >= 8;
  if (filter !== void 0 && include === void 0) {
    console.warn(
      "[vite-plugin-babel] `filter` option is applied after `include` (default since 1.7.0: `/\\.jsx?$/`) and planned for deprecation. Set `include` explicitly to match your filter, or migrate to `include` / `exclude` entirely."
    );
  }
  if (include === void 0) {
    include = DEFAULT_INCLUDE;
  }
  const customFilter = vite.createFilter(include, exclude);
  const transformFilter = (id) => customFilter(id) && testFilter(filter, id);
  const { getBabelOptions, updateRoot } = useBabelConfig(babelConfig);
  const transform = async (code, id) => {
    if (!transformFilter(id)) return;
    const babelOptions = getBabelOptions();
    return babel.transformAsync(code, { ...babelOptions, filename: id }).then((result) => ({ code: result?.code ?? "", map: result?.map }));
  };
  const getOptimizeDeps = () => {
    if (isVite8OrHigher) {
      return {
        rolldownOptions: {
          plugins: [{ name: "rolldown-plugin-babel", transform }]
        }
      };
    }
    return {
      esbuildOptions: {
        plugins: [
          esbuildPluginBabel({
            config: { ...babelConfig },
            transformFilter,
            loader
          })
        ]
      }
    };
  };
  return {
    name: "babel-plugin",
    apply,
    enforce,
    config() {
      return {
        optimizeDeps: getOptimizeDeps(),
        ssr: optimizeOnSSR ? { optimizeDeps: getOptimizeDeps() } : void 0
      };
    },
    configResolved(config) {
      updateRoot(config.root);
    },
    transform
  };
};
function useBabelConfig(babelConfig) {
  let root;
  let babelPartialConfig;
  const getBabelOptions = () => {
    if (babelPartialConfig) return babelPartialConfig.options;
    babelPartialConfig = babel.loadPartialConfig({
      ...babelConfig,
      cwd: root,
      root,
      babelrc: false,
      caller: {
        name: "vite-plugin-babel",
        supportsStaticESM: true,
        ...babelConfig.caller
      }
    });
    return babelPartialConfig?.options ?? {};
  };
  return {
    getBabelOptions,
    updateRoot(newRoot) {
      root = newRoot;
    }
  };
}

exports.default = babelPlugin;
exports.esbuildPluginBabel = esbuildPluginBabel;
